pkg/objectcache/applicationprofilecache/applicationprofilecache.go

```diff
@@ -55,91 +36,252 @@ type ContainerCallStackIndex struct {
+			workloadIDInUse := false
+			for containerID := range containerSet.Iter() {
+				if containerInfo, exists := apc.containerIDToInfo.Load(containerID); exists &&
+					containerInfo.WorkloadID == workloadID {
+					workloadIDInUse = true
+					break
+				}
+			}
+
+			if !workloadIDInUse {
+				continue
+			}
+
+			// Update the profile in the cache
+			if existingProfile, exists := apc.workloadIDToProfile.Load(workloadID); exists {
+				// If the profile already exists and it's complete/completed, continue to the next one
+				if existingProfile.Annotations[helpersv1.CompletionMetadataKey] == helpersv1.Complete {
+					continue
+				}
+
+				// If the new profile is not complete and we already have a completed/partial one, skip it
+				if profile.Annotations[helpersv1.CompletionMetadataKey] != helpersv1.Complete {
+					continue
+				}
+			}
+
+			// Fetch the profile from storage
+			fullProfile, err := apc.storageClient.ApplicationProfiles(namespace).Get(ctx, profile.Name, metav1.GetOptions{})
+			if err != nil {
+				logger.L().Error("failed to get application profile",
+					helpers.String("workloadID", workloadID),
+					helpers.String("namespace", namespace),
+					helpers.Error(err))
+				continue
+			}
+
+			apc.workloadIDToProfile.Set(workloadID, fullProfile)
+			logger.L().Debug("updated profile in cache",
+				helpers.String("workloadID", workloadID),
+				helpers.String("namespace", namespace),
+				helpers.String("status", profile.Annotations[helpersv1.StatusMetadataKey]),
+				helpers.String("completion", profile.Annotations[helpersv1.CompletionMetadataKey]))
+
+			// Update call stack search trees for containers using this workload ID
+			for containerID := range containerSet.Iter() {
+				if containerInfo, exists := apc.containerIDToInfo.Load(containerID); exists &&
+					containerInfo.WorkloadID == workloadID {
+					// Create or update call stack search tree if not exists
+					if _, exists := apc.containerToCallStackIndex.Load(containerID); !exists {
+						// Index the call stacks for this container
+						apc.indexContainerCallStacks(containerID, containerInfo.ContainerID, fullProfile)
+					}
+				}
+			}
+		}
+		return true // Continue to next namespace
+	})
+}
+
+// handleUserManagedProfile handles user-managed profiles
+func (apc *ApplicationProfileCacheImpl) handleUserManagedProfile(profile *v1beta1.ApplicationProfile) {
+	normalizedProfileName := strings.TrimPrefix(profile.Name, "ug-")
+	userManagedProfileUniqueIdentifier := profile.ResourceVersion + string(profile.UID)
+
+	// Create a unique tracking key for this user profile
+	profileKey := fmt.Sprintf("%s/%s", profile.Namespace, normalizedProfileName)
+
+	// Check if we've already processed this exact version of the user-managed profile
+	if storedIdentifier, exists := apc.profileToUserManagedIdentifier.Load(profileKey); exists &&
+		storedIdentifier == userManagedProfileUniqueIdentifier {
+		return
+	}
+
```

Comment:
Can we have more than one user-managed profile for a given workload? I don't know if it works.

Reply:
We can't have more than one.